internal/util/fetch/fetch.go

```diff
@@ -127,142 +129,46 @@ func cloneAndCopy(ctx context.Context, r *git.RepoSpec, dest string) error {
 	if err != nil {
-		return errors.E(op, errors.Git, err)
+		return errors.E(op, errors.Git, errors.Repo(repoSpec.CloneSpec()), err)
 	}
 
-	// clone the repo to a tmp directory.
-	// delete the tmp directory later.
-	err = clonerUsingGitExec(ctx, repoSpec)
-	if err != nil && originalRef != repoSpec.Ref {
-		repoSpec.Ref = originalRef
-		err = clonerUsingGitExec(ctx, repoSpec)
+	packageRef := path.Join(strings.TrimLeft(repoSpec.Path, "/"), repoSpec.Ref)
+	if _, found := upGitRunner.ResolveTag(packageRef); found {
+		repoSpec.Ref = packageRef
 	}
 
+	dir, err := upGitRunner.GetRepo(ctx, []string{repoSpec.Ref})
 	if err != nil {
-		if strings.HasPrefix(repoSpec.Path, "blob/") {
-			p := printer.FromContextOrDie(ctx)
-			p.Printf("git repo contains /blob/, you may need to remove /blob/%s", defaultRef)
-			return errors.E(op, errors.Git, err)
-		}
-		return errors.E(op, errors.Git, err)
+		return errors.E(op, errors.Git, errors.Repo(repoSpec.CloneSpec()), err)
 	}
 
-	return nil
-}
-
-// clonerUsingGitExec is the implementation for cloning a repo from git into
-// a local temp directory. This is used by the public ClonerUsingGitExec
-// function to allow trying multiple different refs.
-func clonerUsingGitExec(ctx context.Context, repoSpec *git.RepoSpec) error {
-	const op errors.Op = "fetch.clonerUsingGitExec"
-	var err error
-	repoSpec.Dir, err = ioutil.TempDir("", "kpt-get-")
-	if err != nil {
-		return errors.E(op, errors.Internal, fmt.Errorf("error creating temp directory: %w", err))
-	}
-	err = runGitExec(ctx, repoSpec.Dir, "init", repoSpec.Dir)
+	gitRunner, err := gitutil.NewLocalGitRunner(dir)
 	if err != nil {
-		return errors.E(op, errors.Git, fmt.Errorf("trouble initializing empty git repo in %s: %w",
-			repoSpec.Dir, err))
+		return errors.E(op, errors.Git, errors.Repo(repoSpec.CloneSpec()), err)
 	}
 
-	err = runGitExec(ctx, repoSpec.Dir, "remote", "add", "origin", repoSpec.CloneSpec())
-	if err != nil {
-		return errors.E(op, errors.Git, fmt.Errorf("error adding remote %s: %w", repoSpec.CloneSpec(), err))
-	}
-	if repoSpec.Ref == "" {
-		repoSpec.Ref, err = gitutil.DefaultRef(repoSpec.Dir)
-		if err != nil {
-			return errors.E(op, errors.Git, fmt.Errorf("error looking up default branch for repo: %w", err))
-		}
+	commit, found := upGitRunner.ResolveRef(repoSpec.Ref)
+	if !found {
+		commit = repoSpec.Ref
 	}
 
-	err = func() error {
-		err = runGitExec(ctx, repoSpec.Dir, "fetch", "origin", "--depth=1", repoSpec.Ref)
-		if err != nil {
-			return errors.E(op, errors.Git, fmt.Errorf("trouble fetching %s, "+
-				"please run 'git clone <REPO>; stat <DIR/SUBDIR>' to verify credentials: %w", repoSpec.Ref, err))
-		}
-
-		err = runGitExec(ctx, repoSpec.Dir, "reset", "--hard", "FETCH_HEAD")
-		if err != nil {
-			return errors.E(op, errors.Git,
-				fmt.Errorf("trouble hard resetting empty repository to %s: %w", repoSpec.Ref, err))
-		}
-		return nil
-	}()
+	_, err = gitRunner.Run(ctx, "reset", "--hard", commit)
```

Comment:
can you add comments to any non-obvious section of the code. For example, why are you running `reset --hard`.

Reply:
I have added more comments about what we are doing here.